src/Highlighter.js

```diff
@@ -18,190 +18,153 @@ const _generateComponentKey = (usedKeys, id) => {
+  const resultComponents = []
+  const componentKeys = {}
+  let components = resultComponents
+  let substringComponentKey = 0
+
+  const textIteratee = (substr, index, indexInPlainText) => {
+    // check whether the caret element has to be inserted inside the current plain substring
+    if (
+      isNumber(caretPositionInMarkup) &&
+      caretPositionInMarkup >= index &&
+      caretPositionInMarkup <= index + substr.length
+    ) {
+      // if yes, split substr at the caret position and insert the caret component
+      const splitIndex = caretPositionInMarkup - index
       components.push(
-        this.getMentionComponentForMatch(id, display, mentionChildIndex, key)
+        renderSubstring(substr.substring(0, splitIndex), substringComponentKey)
       )
+      // add all following substrings and mention components as children of the caret component
+      components = [
+        renderSubstring(substr.substring(splitIndex), substringComponentKey),
+      ]
+    } else {
+      components.push(renderSubstring(substr, substringComponentKey))
     }
 
-    iterateMentionsMarkup(value, config, mentionIteratee, textIteratee)
-
-    // append a span containing a space, to ensure the last text line has the correct height
-    components.push(' ')
-
-    if (components !== resultComponents) {
-      // if a caret component is to be rendered, add all components that followed as its children
-      resultComponents.push(this.renderHighlighterCaret(components))
-    }
+    substringComponentKey++
+  }
 
-    return (
-      <div {...style} ref={containerRef}>
-        {resultComponents}
-      </div>
+  const mentionIteratee = (
+    markup,
+    index,
+    indexInPlainText,
+    id,
+    display,
+    mentionChildIndex,
+    lastMentionEndIndex
+  ) => {
+    const key = _generateComponentKey(componentKeys, id)
+    components.push(
+      getMentionComponentForMatch(id, display, mentionChildIndex, key)
     )
   }
 
-  renderSubstring(string, key) {
+  const renderSubstring = (string, key) => {
     // set substring span to hidden, so that Emojis are not shown double in Mobile Safari
     return (
-      <span {...this.props.style('substring')} key={key}>
+      <span {...style('substring')} key={key}>
         {string}
       </span>
     )
   }
 
-  // Returns a clone of the Mention child applicable for the specified type to be rendered inside the highlighter
-  getMentionComponentForMatch(id, display, mentionChildIndex, key) {
+  const getMentionComponentForMatch = (id, display, mentionChildIndex, key) => {
     const props = { id, display, key }
-    const child = Children.toArray(this.props.children)[mentionChildIndex]
+    const child = Children.toArray(children)[mentionChildIndex]
     return React.cloneElement(child, props)
   }
 
-  // Renders an component to be inserted in the highlighter at the current caret position
-  renderHighlighterCaret(children) {
+  const renderHighlighterCaret = (children) => {
```

Comment:
Missing original comment: `// Renders an component to be inserted in the highlighter at the current caret position`